### PR TITLE
Add BLAKE3 hasher to vfs

### PR DESCRIPTION
**Pull request by tylerwilliams**
Add BLAKE3 hasher to vfs

This PR adds the Blake3Hasher and Blake3HashFunction classes to vfs and
makes them available under the flag --digest_function=BLAKE3.

I spent a little time trying to optimize performance more here, and am happy to
chat about different approaches (ByteBuffer vs byte[] vs unsafe vs mmap). In the
end this CL implements something pretty conservative, but there is room to get
more performance out of this with a different approach.

This is a partial commit for #18658.

### CODE REVIEW

**Comment by tylerwilliams** [comment]
@coeuvre -- Could use your advice on how to fix:

1. Clang on :ubuntu: 18.04 LTS (OpenJDK 11, gcc 7.5.0) -- If we can disable Ubuntu 18 (it's already EOLd), this should just go away. If not, I think possibly a repository_rule could work here to disable including AVX512 on the old ubuntu18 / gcc versions, but this is more complicated.
2. `//src/test/shell/bazel:bazel_bootstrap_distfile_test` is failing, how is this supposed to interact with dist_http_archive? It seemed like that test should be able to run w/o network enabled -- how do I make it aware of the dist_http_archive sources?

**Comment by coeuvre** [comment]
For 1, yes the test is to verify Bazel can be built without network. You need to add blake3 to `distdir_deps.bzl`.

@meteorcloudy what's your opinion on 2?

File: src/main/java/com/google/devtools/build/lib/vfs/DigestHashFunction.java [hunk 1]
```diff
@@ -82,8 +73,6 @@ private DigestHashFunction(
     checkArgument(!names.isEmpty());
     this.name = names.get(0);
     this.names = names;
-    this.messageDigestPrototype = getMessageDigestInstance();
-    this.messageDigestPrototypeSupportsClone = supportsClone(messageDigestPrototype);
   }
 
   public static DigestHashFunction register(
```
Comment (coeuvre) [inline src/main/java/com/google/devtools/build/lib/vfs/DigestHashFunction.java:78]: We cannot change how the registration works here because we rely on this internally. You could, however, glue BLAKE3 into this with following steps:

1. Write a class that implements `java.security.MessageDigest` for `BLAKE3`. This is the class that interact with JNI and call into c code.
2. Register the `MessageDigest` from step 1 by implementing a `java.security.Provider` and add it with `java.security.Security.addProvider(...)`. So that when you call `MessageDigest.getInstance("BLAKE3")`, you can get back the instance from the provider.
3. Adapt `DigestMessage` for `BLAKE3` into `HashFunction` by calling [MessageDigestHashFunction](https://cs.opensource.google/guava/guava/+/master:guava/src/com/google/common/hash/MessageDigestHashFunction.java;l=47;drc=e62d6a0456420d295089a9c319b7593a3eae4a83).
4. Create a `DigestHashFunction` by calling ` DigestHashFunction.register(...)` using the `HashFunction` from step 3.

I know this is verbose and not fun but we have to write it in this way.

**Comment by meteorcloudy** [comment]
For the `bazel_bootstrap_distfile_test`, the issue is subtle:

https://github.com/bazelbuild/bazel/blob/75b02216b30d3c3b4454cedb94464f10c21b768e/distdir_deps.bzl#L276-L281

`"archive": "v1.3.3.zip", ` should be `"archive": "1.3.3.zip", ` to match the URL so that it actually got used during the bootstrap.

**Comment by meteorcloudy** [comment]
For "Clang on :ubuntu: 18.04 LTS (OpenJDK 11, gcc 7.5.0)", we do plan to phase out ubuntu 1804 soon. But we need to schedule the work and ideally we do that for Bazel 7 not in the middle of a LTS release since we promises backwards compatibility. So if there is other ways to workaround it, that will be preferred.

**Comment by tylerwilliams** [comment]
PTAL

@coeuvre -- I've made the changes you suggested and made the implementation a MessageDigest as well as reverting the changes to Fingerprint etc. Thank you for the detailed pointers there. I have https://github.com/bazelbuild/bazel-central-registry/pull/698 out to add BLAKE3 to bcr, which once merged, will allow fixing the remaining CI failures which I think are down to bzlmod.

File: src/main/java/com/google/devtools/build/lib/vfs/Blake3Provider.java [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package com.google.devtools.build.lib.vfs;
+
+import java.security.Provider;
+
+public final class Blake3Provider extends Provider {
+  public Blake3Provider() {
+    super("BLAKE3Provider", "1.0", "A BLAKE3 digest provider");
+    put("MessageDigest.BLAKE3", "com.google.devtools.build.lib.vfs.Blake3MessageDigest");
```
Comment (coeuvre) [inline src/main/java/com/google/devtools/build/lib/vfs/Blake3Provider.java:8]: nit: Use `Blake3MessageDigest.class.getName()` to make refactor easier in the future.

File: src/main/java/com/google/devtools/build/lib/vfs/DigestHashFunction.java [hunk 3]
```diff
@@ -67,6 +68,7 @@ public int getDigestMaximumLength() {
 
   public static final DigestHashFunction SHA1 = register(Hashing.sha1(), "SHA-1", "SHA1");
   public static final DigestHashFunction SHA256 = register(Hashing.sha256(), "SHA-256", "SHA256");
+  public static final DigestHashFunction BLAKE3 = register(new Blake3HashFunction(), "BLAKE3");
```
Comment (coeuvre) [inline src/main/java/com/google/devtools/build/lib/vfs/DigestHashFunction.java:71]: I think you can just use `new MessageDigestHashFunction("BLAKE3")` ([doc](https://cs.opensource.google/guava/guava/+/master:guava/src/com/google/common/hash/MessageDigestHashFunction.java;l=38;drc=e62d6a0456420d295089a9c319b7593a3eae4a83)) instead of implementing `HashFunction` by yourself in `Blake3HashFunction`.

File: src/main/java/com/google/devtools/build/lib/vfs/Blake3MessageDigest.java [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+import com.google.common.hash.Funnel;
+import com.google.common.hash.HashCode;
+import com.google.common.hash.Hasher;
+import com.google.errorprone.annotations.CanIgnoreReturnValue;
+import java.nio.ByteBuffer;
+import java.nio.charset.Charset;
+import java.security.DigestException;
+import java.security.MessageDigest;
+
+public final class Blake3MessageDigest extends MessageDigest implements Hasher {
```
Comment (coeuvre) [inline src/main/java/com/google/devtools/build/lib/vfs/Blake3MessageDigest.java:14]: If you use `MessageDigestHashFunction` as suggested below, you can remove `implements Hasher`.

File: src/main/java/com/google/devtools/build/lib/vfs/Blake3JNI.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.devtools.build.lib.vfs;
+
+import com.google.devtools.build.lib.jni.JniLoader;
+
+final class Blake3JNI {
```
Comment (coeuvre) [inline src/main/java/com/google/devtools/build/lib/vfs/Blake3JNI.java:19]: Please move all Blake3 related classes into their own package since they are specific to Bazel. The `vfs` package is shared with `Blaze` and including Blaze3 classes here will make it hard to import and maintain.

One option is `src/main/java/com/google/devtools/build/lib/vfs/bazel` but it's all up to you.

File: src/main/java/com/google/devtools/build/lib/vfs/DigestHashFunction.java (hunk 3 above)
Comment (coeuvre) [reply src/main/java/com/google/devtools/build/lib/vfs/DigestHashFunction.java:71]: As mentioned above, we don't want put BLAKE3 code in the common code path. Neither the registration here nor the provider below.

I suggest we have a utility class `BazelHashFunctions` which installs the provider and register the hash function, e.g.:

```
class BazelHashFunctions {
    static {
        addProvider(...);
    }

    public static final DigestHashFunction BLAKE3 = DigestHashFunction.register(...);
}
```

In a following PR when we want to create a file system based on BLAKE3, we could do that in `BazelFileSystemModule` with code like:

```
new UnixFileSystem(BazelHashFunctions.BLAKE3, ...);
```

File: src/main/java/com/google/devtools/build/lib/vfs/Blake3MessageDigest.java [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+  // https://github.com/BLAKE3-team/BLAKE3/blob/master/c/blake3.h
+  public static final int KEY_LEN = 32;
+  public static final int OUT_LEN = 32;
+
+  // To reduce the number of calls made via JNI, buffer up to this many bytes.
+  // If a call to "hash()" is made and less than this much data has been
+  // written, a single JNI call will be made that initializes, hashes, and
+  // cleans up the hasher, rather than making separate calls for each operation.
+  public static final int ONESHOT_THRESHOLD = 8 * 1024;
+  private static ThreadLocal<ByteBuffer> threadLocalBuffer = new ThreadLocal<ByteBuffer>();
```
Comment (coeuvre) [inline src/main/java/com/google/devtools/build/lib/vfs/Blake3MessageDigest.java:25]: Why use thread local? I think it's possible to have two instances of this class running on the same thread, or the same instance could be used across threads.

I think creating a buffer per instance should be good enough.

**Review by coeuvre — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/devtools/build/lib/vfs/Blake3MessageDigest.java (hunk 6 above)
Comment (tylerwilliams) [reply src/main/java/com/google/devtools/build/lib/vfs/Blake3MessageDigest.java]: Sure, fine for now. Can revisit performance more later.

File: src/main/java/com/google/devtools/build/lib/vfs/DigestHashFunction.java (hunk 3 above)
Comment (tylerwilliams) [reply src/main/java/com/google/devtools/build/lib/vfs/DigestHashFunction.java]: I'm trying to read between the lines here and figure out where you want stuff / how you want it to be injected -- let me know what I got wrong.

I moved blake3 stuff into a `bazel` subdir of `vfs`. The provider is statically registered when BazelHashFunctions is loaded by the classloader in `src/main/java/com/google/devtools/build/lib/bazel/Bazel.java`. Re the MessageDigestHashFunction -- addressed above, but that's not public afaict.

File: src/main/native/blake3_jni.cc [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <jni.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "external/blake3/c/blake3.h"
```
Comment (tylerwilliams) [inline src/main/native/blake3_jni.cc:19]: @meteorcloudy Can you show me how to correctly refer to this header when it's coming from bzlmod?

The bzlmod tests are failing with error:
```
src/main/native/blake3_jni.cc:19:10: fatal error: external/blake3/c/blake3.h: No such file or directory
   19 | #include "external/blake3/c/blake3.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
```

because bzlmod is putting this source in a directory like `external/blake3~1.3.3/...` which causes them not to be found.

Thanks!

Comment (fmeum) [reply src/main/native/blake3_jni.cc:19]: All repository roots are also include directories, so this works:

```suggestion
#include "c/blake3.h"
```

Comment (tylerwilliams) [reply src/main/native/blake3_jni.cc]: Ah, amazing, thank you @fmeum!

File: src/main/java/com/google/devtools/build/lib/vfs/bazel/Blake3MessageDigest.java [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+    }
+
+    if (buffer.position() > 0) {
+      Blake3JNI.blake3_hasher_update(hasher, buffer.array(), buffer.position());
+      buffer.clear();
+    }
+  }
+
+  public void engineUpdate(byte[] data, int offset, int length) {
+    while (length > 0) {
```
Comment (coeuvre) [inline src/main/java/com/google/devtools/build/lib/vfs/bazel/Blake3MessageDigest.java:47]: One optimization here is we could avoid copying if the size of `data` is larger than `ONESHOT_THRESHOLD`. But this is totally optional and we could leave that for a separate PR.

File: src/main/java/com/google/devtools/build/lib/vfs/bazel/Blake3MessageDigest.java [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+      }
+    }
+  }
+
+  public void engineUpdate(byte[] data) {
+    engineUpdate(data, 0, data.length);
+  }
+
+  public void engineUpdate(byte b) {
+    engineUpdate(new byte[] {b});
```
Comment (coeuvre) [inline src/main/java/com/google/devtools/build/lib/vfs/bazel/Blake3MessageDigest.java:64]: Optional: avoid creating a new object, put the byte into buffer directly.

File: src/main/java/com/google/devtools/build/lib/vfs/bazel/Blake3MessageDigest.java [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+  @Override
+  protected void finalize() throws Throwable {
+    engineReset();
+  }
+
+  /* The following methods implement the {Hasher} interface. */
+
+  @CanIgnoreReturnValue
+  public Hasher putBytes(ByteBuffer b) {
+    buffer = b;
```
Comment (coeuvre) [inline src/main/java/com/google/devtools/build/lib/vfs/bazel/Blake3MessageDigest.java:139]: I think we need to copy/use the argument `b` now instead of saving it for use later. I believe the contract of the interface is the caller is free to use the passed buffer after the function is returned.

File: src/main/java/com/google/devtools/build/lib/vfs/Blake3MessageDigest.java (hunk 4 above)
Comment (coeuvre) [reply src/main/java/com/google/devtools/build/lib/vfs/Blake3MessageDigest.java]: Sorry, I missed that. Can we then create a separate class to `implement Hasher`? I found it is confusing to combine these two especially when we have `update`, `engineUpdate` and share the underlying buffer.

File: src/main/java/com/google/devtools/build/lib/vfs/DigestHashFunction.java (hunk 3 above)
Comment (coeuvre) [reply src/main/java/com/google/devtools/build/lib/vfs/DigestHashFunction.java]: Yes, we want the provider is statically registered by the classloader when it is loading class `BazelHashFunctions`. But we don't want to have it implementing `BlazeModule` and put it in `src/main/java/com/google/devtools/build/lib/bazel/Bazel.java`.

What I proposed above is we could add the code 

```
new UnixFileSystem(BazelHashFunctions.BLAKE3, ...);
```

to somewhere like [here](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/runtime/BazelFileSystemModule.java;l=51;drc=cde925dba0202a7c66458d9e838708ee535994ff) so that when we want to use BLAKE3, it registers it.

We don't need to do it in this CL as we haven't focused on integrating BLAKE3 into Bazel's VFS. Let's limit the scope of this CL to only adapt BLAKE3 as a DigestHashFunction.

Comment (tylerwilliams) [reply src/main/java/com/google/devtools/build/lib/vfs/DigestHashFunction.java]: Can you give a little more context on what else is required to add the digest function to vfs? I'm a little unclear on that part and it will help me for the next cl. Thanks!

Comment (coeuvre) [reply src/main/java/com/google/devtools/build/lib/vfs/DigestHashFunction.java]: Well, I think I overcomplicated that. The only necessary part is we want to register `BLAKE3` for Bazel only. We can just force loading `BazelHashFunctions`  by adding following code block in `BazelFilesystemModule`

```
class BazelFilesystemmodule {
    static {
        BazelHashFunctions.ensureRegistered();
    }
}
```

where `BazelHashFunctions.ensureRegistered` is just an empty static function whose sole purpose is to let JVM load class `BazelHashFunctions` and run the static block there.

File: src/main/java/com/google/devtools/build/lib/vfs/bazel/Blake3MessageDigest.java (hunk 9 above)
Comment (tylerwilliams) [reply src/main/java/com/google/devtools/build/lib/vfs/bazel/Blake3MessageDigest.java]: Done.

File: src/main/java/com/google/devtools/build/lib/vfs/bazel/Blake3MessageDigest.java (hunk 10 above)
Comment (tylerwilliams) [reply src/main/java/com/google/devtools/build/lib/vfs/bazel/Blake3MessageDigest.java]: Right you are.

File: src/main/java/com/google/devtools/build/lib/vfs/bazel/Blake3MessageDigest.java (hunk 8 above)
Comment (tylerwilliams) [reply src/main/java/com/google/devtools/build/lib/vfs/bazel/Blake3MessageDigest.java:37]: Yeah I was trying to keep this simple for now -- enough complexity already :) Agree on deferring this for a followup (where we measure perf).

File: src/main/java/com/google/devtools/build/lib/vfs/Blake3MessageDigest.java (hunk 4 above)
Comment (tylerwilliams) [reply src/main/java/com/google/devtools/build/lib/vfs/Blake3MessageDigest.java]: Done.

File: src/main/java/com/google/devtools/build/lib/vfs/DigestHashFunction.java (hunk 3 above)
Comment (tylerwilliams) [reply src/main/java/com/google/devtools/build/lib/vfs/DigestHashFunction.java]: Thanks, this is just what I needed. Done.

**Review by coeuvre — COMMENTED** [review]
Thanks, this PR is pretty good now! Just a few more comments and we are ready to go!

**Review by coeuvre — APPROVED** [review]

**Comment by coeuvre** [comment]
Thanks! I will import the change now.

**Comment by brentleyjones** [comment]
@coeuvre does this need the awaiting-merge label, or are you manually importing it?

**Comment by coeuvre** [comment]
I am manually importing as we need some internal changes as well.

File: src/main/java/com/google/devtools/build/lib/vfs/bazel/Blake3MessageDigest.java [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+      throw new DigestException("insufficient space in the output buffer to store the digest");
+    }
+
+    byte[] digestBytes = getOutput(OUT_LEN);
+    System.arraycopy(digestBytes, 0, buf, off, digestBytes.length);
+    return digestBytes.length;
+  }
+
+  @Override
+  protected void finalize() throws Throwable {
```
[the remark this answers is not in the document]
Comment (tylerwilliams) [reply src/main/java/com/google/devtools/build/lib/vfs/bazel/Blake3MessageDigest.java:119]: @coeuvre  Your suggestion is somewhat complicated: is there a precedent / do you have an example? I was unable to find one.

We don't want to pass a byte[] (hasher) for every JNI call instead of the pointer; moving bytes across JNI is not free. So when initializing the hasher we'd need to make one call to JNI that returns a byte[], and then another JNI call to get the address of that object. Do you see a better way to do this? Does Java guarantee that the address of the object on the java heap will remain valid?

I also notice there is precedent for using finalize() which is why I thought this would be alright for now.

lmk what you think!

[the remark this answers is not in the document]
Comment (fmeum) [reply src/main/java/com/google/devtools/build/lib/vfs/bazel/Blake3MessageDigest.java:119]: The modern replacement for `finalize` is [`Cleaner`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/ref/Cleaner.html), which I think would be a good fit here.

The address of a `byte` array would not remain fixed. The `GetPrimitiveArrayCritical` JNI method could be used to limit copies, but I don't know whether a hash function runs quickly enough that this wouldn't hurt overall GC performance.

[the remark this answers is not in the document]
Comment (tylerwilliams) [reply src/main/java/com/google/devtools/build/lib/vfs/bazel/Blake3MessageDigest.java]: Thanks @fmeum -- this is the approach I implemented.

[the remark this answers is not in the document]
Comment (coeuvre) [reply src/main/java/com/google/devtools/build/lib/vfs/bazel/Blake3MessageDigest.java]: @fmeum Thanks for the pointer! I think in this case it is fast enough (we do this internally for our hash function).

@tylerwilliams Thanks for updating the code! Now we use `close()` to free the memory but I think we never call it (please correct me if I am wrong). I thought the reason why we used `finalize` is to avoid manual memory management. On the other hand,  if we use `close()`, we could just free memory there instead of using `Cleaner`.

> Do you see a better way to do this? Does Java guarantee that the address of the object on the java heap will remain valid?

What we do internally is something like:

```
// In Java
class Blake3MessageDigest {
    // The value of `sizeof(blake3_hasher)`. If it doesn't have a well known size, we could query using JNI inside following static block.
    private static int STATE_SIZE = xxx;

    // Cached initial state for a new hasher
    private static byte[] INITIAL_STATE = new byte[STATE_SIZE];

    static {
        nativeInit(INITIAL_STATE);
    }

    private byte[] hasher;

    void prepareNativeState() {
        // effectively, it is `blake3_hasher_init(hasher)`. We cache the result to INITIAL_STATE and do a memcpy to avoid JNI roundtrip.
        System.arraycopy(INITIAL_STATE, 0, hasher, 0, STATE_SIZE);
    }

    void callNativeMethodThatNeedAHasher() {
        // Similar to previous checks we had: `hasher == -1`
        if (needPrepareNativeState) {
            prepareNativeState();
        }
        nativeMethodThatNeedAHasher(hasher);
    }

    private static native void nativeInit(byte[] hasher);
    private static native void nativeMethodThatNeedAHasher(byte[] hasher);
}

// JNI
... void xxx_nativeInit(..., jbyteArray jhasher) {
    blake3_hasher *hasher = (blake3_hasher *)env->GetPrimitiveArrayCritical(jstate, nullptr);
    if (hasher) {
        blake3_hasher_init(hasher);
        env->ReleasePrimitiveArrayCritical(jhasher, hasher, JNI_ABORT);
    }
}

... void xxx_nativeMethodThatNeedAHasher(..., jbyteArray jhasher) {
    blake3_hasher *hasher = (blake3_hasher *)env->GetPrimitiveArrayCritical(jstate, nullptr);
    if (hasher) {
        // Do anything with hasher.
        env->ReleasePrimitiveArrayCritical(jhasher, hasher, JNI_ABORT);
    }
}
```

Note that, when passing byte array to JNI, we just pass a pointer, not copying the entire array. We use `GetPrimitiveArrayCritical` to get the underlying buffer which may do a copy, but most of the time there is no copy.

[the remark this answers is not in the document]
Comment (fmeum) [reply src/main/java/com/google/devtools/build/lib/vfs/bazel/Blake3MessageDigest.java]: @couevre As far as I understood the code, `close` is an optional call here, which is a common pattern with `Cleaner`: Immediate release of resources if the caller needs it by calling `close`, guaranteed automated release eventually via the `Cleaner`.

If Blaze successfully uses `GetPrimitiveArrayCritical` internally, it's probably better to use the same pattern here to reduce the risk of performance regressions that go unnoticed because the Bazel code path isn't covered by google3.

[the remark this answers is not in the document]
Comment (tylerwilliams) [reply src/main/java/com/google/devtools/build/lib/vfs/bazel/Blake3MessageDigest.java]: @coeuvre Thanks for the example, this is perfect.

Done.

[the remark this answers is not in the document]
Comment (tylerwilliams) [reply src/main/java/com/google/devtools/build/lib/vfs/bazel/Blake3MessageDigest.java]: Seems like there's now a linking issue due to us doing JNI stuff in static {} -- any ideas on how / why this is happening? I consolidated the JNI code into Blake3MessageDigest so that I could ensure it was running in order and JNI was being loaded before use, but still seeing this.

https://storage.googleapis.com/bazel-untrusted-buildkite-artifacts/01896762-aba0-4951-9be6-b3568825b687/src/test/shell/bazel/bazel_bootstrap_distfile_tar_test/test.log

File: third_party/blake3/blake3.BUILD [hunk 12]
```diff
@@ -30,7 +30,9 @@ cc_library(
     ] + select({
         "@bazel_tools//src/conditions:linux_x86_64": [
             "c/blake3_avx2_x86-64_unix.S",
-            "c/blake3_avx512_x86-64_unix.S",
+            # Disable to appease bazel-ci which uses ubuntu-18 (EOL) and GCC 7
+            # lacking the headers to compile AVX512.
+            # "c/blake3_avx512_x86-64_unix.S",
```
Comment (fmeum) [inline third_party/blake3/blake3.BUILD:35]: @meteorcloudy FYI, this is an unfortunate consequence of the old versions running in CI.

File: src/main/native/blake3_jni.cc [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+
+jbyte *get_byte_array(JNIEnv *env, jbyteArray java_array) {
+  return env->GetByteArrayElements(java_array, nullptr);
+}
+
+void release_byte_array(JNIEnv *env, jbyteArray array, jbyte *addr) {
+  return env->ReleaseByteArrayElements(array, addr, 0);
+}
+
+extern "C" JNIEXPORT long JNICALL
```
Comment (fmeum) [inline src/main/native/blake3_jni.cc:31]: This returns `long` while the corresponding Java method returns `int`. It should also the `JNIEnv *` and `jobject` parameters.

File: src/main/java/com/google/devtools/build/lib/vfs/bazel/Blake3MessageDigest.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (fmeum) [reply src/main/java/com/google/devtools/build/lib/vfs/bazel/Blake3MessageDigest.java]: That usage is fine as it comes after the `loadJni` call. The failing test bootstraps Bazel and may not honor BUILD files, maybe the new JNI sources need to be listed somewhere else?

File: src/main/native/blake3_jni.cc (hunk 13 above)
Comment (tylerwilliams) [reply src/main/native/blake3_jni.cc:31]: Done.

File: src/main/java/com/google/devtools/build/lib/vfs/bazel/Blake3MessageDigest.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (tylerwilliams) [reply src/main/java/com/google/devtools/build/lib/vfs/bazel/Blake3MessageDigest.java]: @fmeum -- that was my intuition too but I can't figure out where to add these. `blake3_jni.cc` should already be a part of the `src/main/native:srcs` filegroup afaict?

```
filegroup(
    name = "srcs",
    srcs = glob(["**"]) + [
        "//src/main/native/windows:srcs",
    ],
    visibility = ["//src:__pkg__"],
)
```

[the remark this answers is not in the document]
Comment (coeuvre) [reply src/main/java/com/google/devtools/build/lib/vfs/bazel/Blake3MessageDigest.java]: I think you need to trigger JNI loader similar to https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/unix/NativePosixSystem.java;l=28;drc=50adefb987351eb3612b2e28240cf0d0697fdf5a.

[the remark this answers is not in the document]
Comment (coeuvre) [reply src/main/java/com/google/devtools/build/lib/vfs/bazel/Blake3MessageDigest.java]: Oops, nevermind, you already did that. Let me dig more.

File: third_party/blake3/blake3.BUILD (hunk 12 above)
Comment (brentleyjones) [reply third_party/blake3/blake3.BUILD:35]: Related: https://github.com/bazelbuild/bazel/pull/18784#issuecomment-1609421453

File: src/main/java/com/google/devtools/build/lib/vfs/bazel/Blake3MessageDigest.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (coeuvre) [reply src/main/java/com/google/devtools/build/lib/vfs/bazel/Blake3MessageDigest.java]: Okay, I found the reason. For the bootstrap test, we compile stage1 Bazel using [compile.sh](https://cs.opensource.google/bazel/bazel/+/master:scripts/bootstrap/compile.sh;l=408;drc=30b21303b0fd65c70285ea0832db18f78ad9fd0f) where we don't build JNI but fallback to use NIO. We build a stage2 Bazel using the binary from stage1, which doesn't have the JNI, so we failed to initialize Blake3MessageDigest.

How it works for `UnixFileSystem` is, we [detect whether JNI is available](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/bazel/BazelFileSystemModule.java;l=75;drc=6d6bbdcd7ac7e8130741417d891e641495b046b2). If not fallback to use NIO based file system.

In order to pass the test, we should do something similar in `BazelHashFunctions` where we only register Blake3 if JNI is available.

[the remark this answers is not in the document]
Comment (tylerwilliams) [reply src/main/java/com/google/devtools/build/lib/vfs/bazel/Blake3MessageDigest.java]: Done.

**Comment by tylerwilliams** [comment]
@coeuvre I believe this is ready for another look / attempt at merging. Thanks

**Comment by coeuvre** [comment]
Thanks, I am importing now!

**Comment by tylerwilliams** [comment]
> Everything went well. Currently waiting for internal review.

@coeuvre sweet, thanks! Btw did you see my comment on the gerrit repo? https://bazel-review.googlesource.com/c/bazel/+/225070

Hoping we can make that last minute change before it goes in. Thanks again!

**Comment by coeuvre** [comment]
Sorry, I didn't see that because there wasn't any notification :(.

Np, I can make the change now.

**Comment by brentleyjones** [comment]
Thanks @coeuvre!

**Comment by iancha1992** [comment]
The changes in this PR have been included in [Bazel 6.4.0 RC1](https://releases.bazel.build/6.4.0/rc1/index.html). Please test out the release candidate and report any issues as soon as possible. If you're using Bazelisk, you can point to the latest RC by setting USE_BAZEL_VERSION=last_rc.
Thanks!
